Interpolate I8 nearest planar

apavliuk-altran opened the pull request:
### Details:
 - update for Interpolate nearest planar kernel to support int8/uint8 precision without fusing and type conversion; is enabled for AVX2 and SSE41 instruction sets.
 - update for Interpolate nearest planar kernel to support fp32 precision with fusing and type conversion; is enabled for SSE41 instruction set
 - new Interpolate CPU function tests which cover these kernels

### Tickets:
 - [CVS-55387]

apavliuk-altran commented:
@lohika-denis-kotov, please take a look

dmitry-gorokhov commented:
@chenhu-wang Please take a look

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -272,7 +277,107 @@ struct jit_uni_interpolate_kernel_f32 : public jit_uni_interpolate_kernel, publi
+     * @param byte_offset Offset of the current byte [0..3]
+     * @return none.
+     */
+    inline void emulate_gather_byte_impl(const Xbyak::Xmm &xmm_arg, const Xbyak::Reg64 &mem_base, const Xmm &xmm_index,
+                                         const Xbyak::Reg64 &mem_aux, int byte_offset) {
+        constexpr int xmm_len = cpu_isa_traits<sse41>::vlen;
+        const size_t src_prc_size = jcp_.src_prc.size();
+        if (src_prc_size != 1) {
+            IE_THROW() << "The data type of size '" << src_prc_size << "' is not supported.";
+        }
+        constexpr int ind_block_size = xmm_len / sizeof(int32_t);
+        for (int i = 0; i < ind_block_size; ++i) {
+            vpextrd(Xbyak::Reg32(mem_aux.getIdx()), xmm_index, i);
+            add(mem_aux, mem_base);
+            vpinsrb(xmm_arg, xmm_arg, ptr[mem_aux], i + byte_offset * ind_block_size);
+        }
+    }
```
Are you able to emulate with simple mov() based on top of stack pointed by rsp. _extrd() and _insrb() have relative larger and varied latency on different platforms. We have similar gather emulation implementation in interpolate cubic mode. You can ref here:https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/src/nodes/interpolate.cpp#L1286-L1309

chenhu-wang replied:
Use uni_xxx(such as uni_vpinsrb()) version as much as possible.

1. Run vpinsrb() on sse41 will have illegal instruction error.

2. Run pinsrb() on avx2/avx512 have performance penalty sometimes. 

use uni_xxx can help you handle this.

apavliuk-altran replied:
@chenhu-wang, Changed gather as you suggested, it gained some performance.

dmitry-gorokhov commented:
@chenhu-wang Do you have any additional comments?

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -1317,6 +1338,35 @@ struct jit_uni_interpolate_kernel_f32 : public jit_uni_interpolate_kernel, publi
+     * @param is_scalar Flag to perform scalar operation instead of vector
+     * @return none.
+     */
+    inline void gather_i8_i32_indices_store(const Xbyak::Reg64 &dest_addr, const Xbyak::Reg64 &source_addr,
+                                            const Xbyak::Reg64 &ind_addr, bool is_scalar) {
```
```suggestion
    inline void gather_i8_i32_indices_store(const Xbyak::Reg64 &dest_addr, const Xbyak::Reg64 &source_addr,
                                            const Xbyak::Reg64 &ind_addr, int gather_num) {
```
let support this signature to optimize tails process.

apavliuk-altran replied:
Done

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -361,11 +375,18 @@ struct jit_uni_interpolate_kernel_f32 : public jit_uni_interpolate_kernel, publi
-                    apply_post_ops(jcp_.dst_prc, 1);
-                store(vmm_val, reg_dst, scalar_step);
-
+                if (is_i8) {
+                    load(reg_src_aux, vmm_val, scalar_step, 0, Precision::I8);
+                    if (attr_.post_ops_.len() != 0) {
+                        IE_THROW() << "Interpolate I8 doesn't support fusing";
+                    }
+                    store(vmm_val, reg_dst, scalar_step, 0, Precision::I8);
+                } else {
+                    load(reg_src_aux, vmm_val, scalar_step);
+                    if (attr_.post_ops_.len() != 0)
+                        apply_post_ops(jcp_.dst_prc, 1);
+                    store(vmm_val, reg_dst, scalar_step);
+                }
```
for all the tails, we can handle it in one vector process, without loop. Gather all the tails to vector register + post_ops + store with store_number parameter. To support this, we need emu_gather() with gather_num parameter.

apavliuk-altran replied:
Done

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -296,6 +298,9 @@ struct jit_uni_interpolate_kernel_f32 : public jit_uni_interpolate_kernel, publi
     }
 
     void nn_planar() {
+        const bool is_i8 = jcp_.src_prc == InferenceEngine::Precision::I8
+                           && jcp_.dst_prc == InferenceEngine::Precision::I8;
+
```
I believe we can support u8 as well with current implementation. Just relax conditions.

apavliuk-altran replied:
Done

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -327,26 +333,34 @@ struct jit_uni_interpolate_kernel_f32 : public jit_uni_interpolate_kernel, publi
                 jl(nn_loop_end_label, T_NEAR);
+                if (is_i8) {
+                    gather_i8_i32_indices_store(reg_dst, reg_src_h, reg_index, false);
+                    if (attr_.post_ops_.len() != 0) {
+                        IE_THROW() << "Interpolate I8 doesn't support fusing";
+                    }
```
Report exceptions will just lead some models fail to run, we need manually handle it to run successfully. Two options:

1. Disable fusion directly for this case in [canBeFuse()](https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/src/nodes/interpolate.cpp#L3228). Then there will be no post-ops in this case, and post-ops is processed in another node successfully.
2. Support fusion in all cases. If there are post-ops, we need gather u8/i8(src_prc) data to FP32 in register(need precision convert), apply post-ops fusion, store FP32 to dst_prc with store_num. In this case, vector step is always (vlen / sizeof(float));

@dmitry-gorokhov , @ceciliapeng2011 what do you think which one is preferred?

apavliuk-altran replied:
For now, added a check to `Interpolate::canFuse()`

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -1317,6 +1338,35 @@ struct jit_uni_interpolate_kernel_f32 : public jit_uni_interpolate_kernel, publi
+    inline void gather_i8_i32_indices_store(const Xbyak::Reg64 &dest_addr, const Xbyak::Reg64 &source_addr,
+                                            const Xbyak::Reg64 &ind_addr, bool is_scalar) {
+        constexpr int gpr_size = 8;
+        sub(rsp, gpr_size);
+        // move content in register to content in address(ptr[])
+        mov(ptr[rsp], reg_tmp_64);
```
Let remove these lines, I believe it's safe in current code path.

apavliuk-altran replied:
Done

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -1317,6 +1338,35 @@ struct jit_uni_interpolate_kernel_f32 : public jit_uni_interpolate_kernel, publi
+            mov(ptr[dest_addr + i], reg_tmp_64.cvt8());
+        }
+
+        // restore GPR state
+        mov(reg_tmp_64, ptr[rsp]);
+        add(rsp, gpr_size);
```
Let remove this 2 lines.

apavliuk-altran replied:
Done

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -1771,9 +1821,14 @@ void Interpolate::initSupportedPrimitiveDescriptors() {
-        if (mayiuse(cpu::x64::avx2) && inputPrecision == Precision::FP32) {
+        const bool allowAvx2NcspFp32 = mayiuse(cpu::x64::avx2) && inputPrecision == Precision::FP32;
+        // allow I8 planar format only for avx2 with channels = 1, fusing is not supported
+        const bool allowAvx2NcspI8 = mayiuse(cpu::x64::avx2) && !mayiuse(cpu::x64::avx512_core) &&
```
why not support sse41 and avx512 target, the emulation should and does not have such limitations.

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -1771,9 +1821,14 @@ void Interpolate::initSupportedPrimitiveDescriptors() {
+        const bool allowAvx2NcspFp32 = mayiuse(cpu::x64::avx2) && inputPrecision == Precision::FP32;
+        // allow I8 planar format only for avx2 with channels = 1, fusing is not supported
+        const bool allowAvx2NcspI8 = mayiuse(cpu::x64::avx2) && !mayiuse(cpu::x64::avx512_core) &&
+            inputPrecision == Precision::I8 && outputPrecision == Precision::I8 &&
```
extend to U8 as well.

apavliuk-altran replied:
Done

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -1875,10 +1930,15 @@ void Interpolate::prepareParams() {
         std::shared_ptr<InterpolateExecutor> executor;
+        const bool isPlanar = key.nodeAttrs.layout == InterpolateLayoutType::planar;
+        // allow I8 planar format only for avx2 with channels = 1, fusing is not supported
+        const bool allowNearestPlanarI8 = isPlanar && mayiuse(cpu::x64::avx2) && !mayiuse(cpu::x64::avx512_core) &&
+            key.nodeAttrs.inPrc == Precision::I8 && key.nodeAttrs.outPrc == Precision::I8 &&
+            key.srcDims[1] == 1 && key.attr.get_post_ops().len() == 0;
```
could you please extend ISA and precisions.

apavliuk-altran replied:
Done

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -3118,17 +3178,22 @@ Interpolate::InterpolateJitExecutor::InterpolateJitExecutor(const InterpolateAtt
-    } else if (mayiuse(cpu::x64::avx2) && interpAttrs.inPrc == InferenceEngine::Precision::FP32) {
-        // gather ISA(for planar JIT kernel) for avx2 and fp32
-        interpolateKernel.reset(new jit_uni_interpolate_kernel_f32<cpu::x64::avx2>(jcp, *attr.get()));
+    } else if (mayiuse(cpu::x64::avx2) && (allowFp32 || allowI8)) {
+        interpolateKernel.reset(new jit_uni_interpolate_kernel_f32_i8<cpu::x64::avx2>(jcp, *attr.get()));
```
```suggestion
    } else if (mayiuse(cpu::x64::avx2) && allowFp32) {
        interpolateKernel.reset(new jit_uni_interpolate_kernel_f32_i8<cpu::x64::avx2>(jcp, *attr.get()));
    } else if (allowI8) {
        if (mayiuse(cpu::x64::avx512_core)) {
            interpolateKernel.reset(new jit_uni_interpolate_kernel_f32_i8<cpu::x64::avx512_core>(jcp, *attr.get()));
        } else if (cpu::x64::avx2) {
            interpolateKernel.reset(new jit_uni_interpolate_kernel_f32_i8<cpu::x64::avx2>(jcp, *attr.get()));
        } else if (cpu::x64::sse41) {
           interpolateKernel.reset(new jit_uni_interpolate_kernel_f32_i8<cpu::x64::sse41>(jcp, *attr.get()));
        }
    }
```

apavliuk-altran replied:
Done

apavliuk-altran commented:
Updated the PR to support U8 precision and all ISA for I8/U8 precisions.
After running the benchmarks for various shapes and Interpolate params, added a Threshold table with thresholds for blocked and channel-first layouts.

apavliuk-altran commented:
Left threshold for channel first and blocked layouts as 1 according to the new benchmarks and suggestion from @chenhu-wang.

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -3281,11 +3289,13 @@ size_t Interpolate::getSpatialDimsNum(const Dim rank) {
-               interpAttrs.inPrc == InferenceEngine::Precision::I8 &&
-               interpAttrs.outPrc == InferenceEngine::Precision::I8;
-    if (!mayiuse(cpu::x64::sse41) || is_i8_nearest_planar || interpAttrs.mode == InterpolateMode::linear) {
+    const bool isNearestPlanar = interpAttrs.mode == InterpolateMode::nearest &&
+                                 interpAttrs.layout == InterpolateLayoutType::planar;
+    const bool isI8OrU8 =
+        (interpAttrs.inPrc == InferenceEngine::Precision::I8 && interpAttrs.outPrc == InferenceEngine::Precision::I8) ||
+        (interpAttrs.inPrc == InferenceEngine::Precision::U8 && interpAttrs.outPrc == InferenceEngine::Precision::U8);
+
+    if (!mayiuse(cpu::x64::sse41) || (isNearestPlanar && isI8OrU8) || interpAttrs.mode == InterpolateMode::linear) {
```
As far as I understand, interpAttrs.layout, interpAttrs.inPrc and interpAttrs.outPrc is not finalized yet at the time when this function is called. If jit kernel does not support fusion, here need an overkilled conditions, interpAttrs.mode == InterpolateMode::nearest  and getOriginalInputPrecisionAtPort(DATA_ID) == I8 or U8, no matter what layout it is. Ideally we should support fusion at all jitted kernels and remove these restriction.

apavliuk-altran replied:
This code is removed as we decided to support fusing for `int8` later

chenhu-wang commented on src/plugins/intel_cpu/src/nodes/interpolate.cpp:
```diff
@@ -1786,35 +1787,46 @@ void Interpolate::initSupportedPrimitiveDescriptors() {
 
-    if (!mayiuse(cpu::x64::sse41) || interpAttrs.mode == InterpolateMode::linear) {
-        pushDesc(LayoutType::ncsp, impl_type);
+    // planar for 1.ref on machine without sse41(if no sse41, canFuse() is false). 2.JIT kernel for f32 && avx2(gather).(with fuse)
```
could you please update code explanation comments to cover planar support by emulation case.

apavliuk-altran replied:
Done

apavliuk-altran commented:
@ceciliapeng2011 , @chenhu-wang , @dmitry-gorokhov 
If you don't have any more comments, could you please merge this PR after all checks will have passed?

ceciliapeng2011 commented:
@iefode and @jiwaszki, can you please take a look at this pr from ie-test and samples maintainers' perspective?

iefode commented on src/plugins/intel_cpu/tests/functional/single_layer_tests/interpolate.cpp:
```diff
@@ -254,10 +254,18 @@ class InterpolateLayerCPUTest : public testing::WithParamInterface<InterpolateLa
         selectedType = makeSelectedTypeStr(selectedType, ngPrc);
+
+        const auto selLength = selectedType.length();
+        if (selLength > 1 && selectedType[selLength - 2] == 'U' && selectedType[selLength - 1] == '8') {
```
quite strange hardcode

iefode replied:
you check that it's `U8` and change the type after that without any reason.
Moreover, the comparation looks so strange (we compare the chars instead of element type). Why do you use such constractions?

apavliuk-altran replied:
@iefode , the reason for this is to be consistent with the following workaround where exactly the same thing is done in existing code.
Could you please take a look at it?
https://github.com/openvinotoolkit/openvino/blob/c81fa39d3d40e305aa3783de2376ca3b3326533b/src/plugins/intel_cpu/src/node.cpp#L436
https://github.com/openvinotoolkit/openvino/blob/c81fa39d3d40e305aa3783de2376ca3b3326533b/src/plugins/intel_cpu/src/node.cpp#L442

apavliuk-altran replied:
@iefode , did you have an opportunity to take a look at the original workaround?

iefode commented on src/plugins/intel_cpu/tests/functional/single_layer_tests/interpolate.cpp:
```diff
@@ -334,12 +348,18 @@ const std::vector<fusingSpecificParams> interpolateFusingParamsSet{
-std::vector<std::map<std::string, std::string>> filterAdditionalConfig() {
+std::vector<std::map<std::string, std::string>> filterAdditionalConfig(bool enforceBF16 = true) {
     if (InferenceEngine::with_cpu_x86_avx512f()) {
+        if (enforceBF16) {
```
look like to better use these configs in instansiation only for bf16

apavliuk-altran replied:
Unfortunately, I haven't fully got your point.
The following key: `{InferenceEngine::PluginConfigParams::KEY_ENFORCE_BF16, InferenceEngine::PluginConfigParams::YES}` was present in original tests, and the only original precision used there was `ElementType::f32`
Do you suggest to completely remove this key until we add full BF16 support and corresponding tests?

iefode replied:
better to separate test instantiations for bf16 to standalone instance according using bf16 configs.

apavliuk-altran replied:
Created separate function `filterAdditionalConfigNearestExt()` for Interpolate Nearest tests.
@iefode , could you please take a look?

akladiev commented:
This PR will be closed in 2 weeks in case of no activity.

akladiev commented:
This PR was closed because it has been stalled for 2 week with no activity.
